File: 20.valid-parentheses.ts
/*
 * @lc app=leetcode id=20 lang=typescript
 *
 * [20] Valid Parentheses
 */

// @lc code=start
function isValid(s: string): boolean {
    const stack: string[] = [];
    for (let i = 0; i < s.length; i++) {
        const c = s[i];
        switch (c) {
            case '(':
            case '[':
            case '{':
                stack.push(c);
                break;
            case ')':
                if (stack.pop() !== '(') return false;
                break;
            case ']':
                if (stack.pop() !== '[') return false;
                break;
            case '}':
                if (stack.pop() !== '{') return false;
                break;
        }
    }
    return stack.length === 0;
};
// @lc code=end

File: 2024/20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        close_map = {'{': '}', '[': ']', '(': ')'}
        for c in s:
            if len(stack) == 0 and c not in close_map:
                return False
            if c in close_map:
                stack.append(c)
            elif close_map[stack[-1]] != c:
                return False
            else:
                stack.pop()
        return len(stack) == 0
# @lc code=end

